Added Kafka as Transport - part 2

vMaroon opened the pull request.

vMaroon committed:
generalized transport interface usage

vMaroon committed:
refactored

vMaroon committed:
committing bundle is now done after processing

vMaroon committed:
refactored

vMaroon committed:
refactored

vMaroon committed:
refactored

vMaroon committed:
- Dockerfile fix to stabilise built binaries
- Added kafka as transport

vMaroon committed:
mods

nirrozenbaum committed:
Merge branch 'main' into kafka

nirrozenbaum commented on README.md:
````diff
@@ -36,18 +36,36 @@ Go to the [Contributing guide](CONTRIBUTING.md) to learn how to get involved.
     $ export REGISTRY=...
     ```
-    
-1.  Set the `IMAGE` environment variable to hold the name of the image.
````
IMAGE was removed by mistake from README

nirrozenbaum commented on pkg/transport/kafka/consumer_configuration.go:
```diff
@@ -0,0 +1,64 @@
+
+var errEnvVarNotFound = errors.New("environment variable not found")
+
+func getKafkaConfig() (*kafka.ConfigMap, []string, error) {
```
change function name to readEnvVars + move file content to consumer.go file (see example in spec-transport-bridge)

nirrozenbaum commented on pkg/transport/kafka/consumer_configuration.go:
```diff
@@ -0,0 +1,64 @@
+
+const (
+	envVarKafkaConsumerID    = "KAFKA_CONSUMER_ID"
+	envVarKafkaConsumerHosts = "KAFKA_BOOTSTRAP_SERVERS"
```
envVarKafkaBootstrapServers

nirrozenbaum commented on pkg/transport/kafka/consumer_configuration.go:
```diff
@@ -0,0 +1,64 @@
+const (
+	envVarKafkaConsumerID    = "KAFKA_CONSUMER_ID"
+	envVarKafkaConsumerHosts = "KAFKA_BOOTSTRAP_SERVERS"
+	envVarKafkaConsumerSSLCA = "KAFKA_SSL_CA"
```
envVarKafkaSSLCA

nirrozenbaum commented on build/Dockerfile:
```diff
@@ -30,5 +52,10 @@ RUN  /usr/local/scripts/user_setup
 RUN microdnf update && \
     microdnf clean all
 
+# allow user to fully control the bin directory (for transport certificates) -- for now
```
fix comment

nirrozenbaum commented on pkg/transport/kafka/consumer_configuration.go:
```diff
@@ -0,0 +1,64 @@
+	envVarKafkaConsumerID    = "KAFKA_CONSUMER_ID"
+	envVarKafkaConsumerHosts = "KAFKA_BOOTSTRAP_SERVERS"
+	envVarKafkaConsumerSSLCA = "KAFKA_SSL_CA"
+	envVarKafkaTopic         = "KAFKA_TOPICS"
```
envVarKafkaTopics (add s)

nirrozenbaum commented on pkg/transport/kafka/kafka_consumer_client.go:
```diff
@@ -0,0 +1,292 @@
+package kafka
```
rename file name to consumer.go (like struct)

nirrozenbaum commented on README.md:
````diff
@@ -36,18 +36,36 @@ Go to the [Contributing guide](CONTRIBUTING.md) to learn how to get involved.
+    $ export TRANSPORT_TYPE=...
     ```
-    $ export IMAGE=$REGISTRY/$(basename $(pwd)):latest
+1. If you set Kafka for transport, set the following environment variables:
````
let's be consistent with mentioning transport options in lowercase letters.
if you set kafka as transport...

nirrozenbaum commented on README.md:
````diff
@@ -36,18 +36,36 @@ Go to the [Contributing guide](CONTRIBUTING.md) to learn how to get involved.
+    $ export KAFKA_SSL_CA=$(cat PATH_TO_CA | base64 -w 0)
     ```
 
+1. Otherwise, if you set Sync-Service as transport, set the following:
````
if you set sync-service (lowercase letters)

nirrozenbaum commented on pkg/transport/kafka/kafka_consumer_client.go:
```diff
@@ -0,0 +1,292 @@
+	}
+
+	if err := kafkaConsumer.Subscribe(topics); err != nil {
+		return nil, fmt.Errorf("failed to create consumer: %w", err)
```
change error to `failed to subscribe to requested topics` + add `kafkaConsumer.Close()` to avoid memory leak.

nirrozenbaum commented on pkg/transport/kafka/kafka_consumer_client.go:
```diff
@@ -0,0 +1,292 @@
+)
+
+// NewConsumer creates a new instance of Consumer.
+func NewConsumer(log logr.Logger, bundleUpdatesChan chan *bundle.ObjectsBundle) (*Consumer, error) {
```
see all comments in this file from previous PR.. just open the file and drill down to see all comments:
https://github.com/open-cluster-management/leaf-hub-spec-sync/pull/6/files

nirrozenbaum reviewed (commented):
@vMaroon I've added some comments.
we can continue together if you prefer.

vMaroon committed:
compliance to review comments

vMaroon committed:
refactored

vMaroon committed:
refactored sync service,
refactored consumer + re-implemented committing logic

vMaroon committed:
funcs order

vMaroon committed:
lock protection on conc map read/writes

vMaroon committed:
remodeled the committing logic:
transport now passes bundles with metadata to the bundle syncer (processor) which gets committed upon successful processing.

nirrozenbaum commented on pkg/transport/sync-service/sync_service.go:
```diff
@@ -98,46 +106,103 @@ type SyncService struct {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	s.objectMetadataToCommitMap[fmt.Sprintf("%s.%s", objectMetadata.ObjectType,
```
in sync service I didn't understand why we need map..
isn't a FIFO queue enough?
CommitAsync - push to end of FIFO
commitObjects func - (pseudo code) 
```
while objectMetadata := commitQueue.Dequque() != nil {
        syncServiceClient.MarkObjectConsumed(objectMetadata)
```
to avoid ugly lock/unlock code, we need to implement FIFO concurrent queue that implements Enquque/Dequque under lock, while dequeue should return nil in case queue is empty.
(something that lock more or less like ready_queue in status bridge, but without blocking).

vMaroon replied:
why commit outdated offsets/metadata? we could wrap the map with a structure with lock protection but that would turn every operation into a lock acquire/release which is less efficient than lock -> execute work -> unlock. we could instead consider sync.Map but I personally think this code is alright, the lock/unlock calls are only avoiding the committing part and are clear enough / hard to confuse.

vMaroon replied:
we could also avoid all this by reverting back to having CommitAsync execute its content in a new go routine, then the routine's job / the committer routine can have a simple lock.lock and defer lock.unlock..

nirrozenbaum commented on pkg/transport/kafka/consumer.go:
```diff
@@ -0,0 +1,312 @@
+	}
+}
+
+func (c *Consumer) commitMappedOffsets() {
+	c.lock.Lock()
+	if len(c.partitionToOffsetToCommitMap) == 0 {
+		c.lock.Unlock()
+		return
+	}
+
+	topicPartitions := make([]kafka.TopicPartition, 0, len(c.partitionToOffsetToCommitMap))
+
+	// prepare batch for committing
+	for partition, highestOffset := range c.partitionToOffsetToCommitMap {
+		topicPartitions = append(topicPartitions, kafka.TopicPartition{
+			Topic:     &c.topic,
+			Partition: partition,
+			Offset:    highestOffset + 1, // kafka re-processes the committed offset on restart, so +1 to avoid that.
+		})
+	}
+
+	c.lock.Unlock()
+
+	if _, err := c.kafkaConsumer.Consumer().CommitOffsets(topicPartitions); err != nil {
+		c.log.Error(err, "failed to commit offsets", "TopicPartitions", topicPartitions)
+		return
+	}
+
+	// update offsets map, delete what's been committed
+	c.lock.Lock()
+	for _, topicPartition := range topicPartitions {
+		if c.partitionToOffsetToCommitMap[topicPartition.Partition] == topicPartition.Offset {
+			// no new offsets processed on this partition, delete from map
+			delete(c.partitionToOffsetToCommitMap, topicPartition.Partition)
+		}
+	}
+	c.lock.Unlock()
+
```
same comment as in sync service..
can we do the same with concurrent queue instead of map?
you can push to the queue TopicPartition pointer.
when we dequeue all, we can arrange the data to commit (calculate highest offset).
then there is no need to do another round of comparing pointers + deleting objects from map.
(theoretically, we could just go over the topic partition and commit all of them, calculating highest in the client side is an optimization)

nirrozenbaum reviewed (commented):
@vMaroon see comments

vMaroon committed:
refactored
